walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,154 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
+
+  /*
+   * Given previously-scheduled events, query() finds a list of all the possible time slots
+   * throughout a single day that the requested meeting can take place. If the request contains
+   * optional attendees, query() will try to find if there is at least one time slot
+   * that both required and optional attendees can attend. If a time slot cannot be found,
+   * only time slots that required attendees can attend will be returned.
+   *
+   * @param events: List of already-scheduled events that may conflict with meeting request.
+   * @param request: Meeting request (duration and a list of attendees) to be scheduled 
+   * into the day
+   * @return a list of the possible time ranges for the meeting request to take place.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    /*
+     * Request duration is longer than a day, which is impossible to find an appropriate
+     * time range for. Returns an empty list.
+     */
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return new ArrayList<TimeRange>();
+    }
+    
+    ArrayList<Event> eventsArr = sortEventsByStart(events);
+
+    if (request.hasOptionalAttendees()) {
+      ArrayList<TimeRange> optionalSlots = findTimeSlots(eventsArr, request, 
+        /* includeOptionalAttendees = */ true);
+      if (optionalSlots.isEmpty() && request.hasAttendees()) {
+        return findTimeSlots(eventsArr, request, /* includeOptionalAttendees = */ false);
+      }
+      return optionalSlots;
+    } else {
+      return findTimeSlots(eventsArr, request, /* includeOptionalAttendees = */ false);
+    }
+  }
+
+  /*
+   * Sort events by the start of their TimeRanges.
+   */
+  private ArrayList<Event> sortEventsByStart(Collection<Event> events) {
+    ArrayList<Event> eventsArr = new ArrayList<>(events);
+    Collections.sort(eventsArr, new Comparator<Event>() {
+      @Override
+      public int compare(Event a, Event b) {
+        return Long.compare(a.getWhen().start(), b.getWhen().start());
+      }
+    });
+    return eventsArr;
+  }
+
+  /*
+   * Finds all the possible time ranges throughout the day that the requested
+   * meeting can take place, given already-scheduled events.
+   * 
+   * @param events: List of already-scheduled events that may conflict with the times
```

Comment:
It seems like changing the parameter from `Collection<Event> events` to `List<Event> events` causes errors when query() is called, even though ArrayLists are always used to call the function. Is it sufficient to specify in the javadoc that the input should be a `List<Event>`?

Reply:
Following up from some Chat messages, I think that the best approach is to avoid sorting this input at all, if sorting the output will work instead, assuming that sorted output is the ultimate purpose of sorting the input. Sorting the output would more directly express the intended behavior and be more robust to future changes that might rearrange the slots as they're being checked.

If sorting the output isn't a working/ideal approach, I think that sorting the input at the beginning of this function, rather than relying on the caller to do it before passing it in, would be a good approach. It makes this method more "fool-proof" to use, which becomes more and more important as code is called from more places, especially if people ignore its documentation that says how the input should be organized.